### DISCUSSION TITLE
Simplified "boolean strategy" handling

### DISCUSSION THREAD

**Discussion by sebersole** [hibernate/hibernate-orm#3966] [Design Proposals]
We have support for storing a boolean in the domain model in multiple ways in the database.  In addition to normal BIT/BOOLEAN storage, we also support Y/N, T/F and 1/0.

At the moment, to use any of these you have to explicitly use `@Type` to specify the mapping.  E.g.

```
@Basic
@Type( type = "yes_no" )
boolean enabled;
```

I've been thinking it would be nice to have a simpler way to specify these.  This might be an annotation

```
@Basic
@Boolean( strategy = YES_NO )
boolean enabled;
```

or maybe just publishing AttributeConverter impls for these cases.

WDYT?

**Comment by dreab8** [comment]
I like `@Boolean( strategy = YES_NO )`

**Comment by beikov** [comment]
I like the idea, but only in addition to the possibility to of using `@type` as because:
* The annotation with enum approach is not extendible (what if some user wants `J`/`N` instead of `Y`/`N`)
* I think we need to handle the types internally in a special way as we do for the `YesNoType`, `TrueFalseType` and `NumericBooleanType`. Not sure if the attribute converter approach works here

So overall, I favor the annotation approach for exposing builtin strategies. Users can still use custom types though.

**Comment by gavinking** [comment]
Dunno. I guess I'm sorta not _that_ keen on introducing an enum to model something that's in principle open-ended. I mean, yes, sure, for legacy reasons and convenience, we have these three built-in options, and that's OK as far as it goes, in that it's a pretty minimal API surface. But it's not like it's such a really super-excellent model that we should be freezing it into new constructs.

(Also I don't think you would literally call it `Boolean`, since you probably want to be able to use it on a field of type `java.lang.Boolean`.)

> maybe just publishing `AttributeConverter` impls for these cases.

That sounds like a better option to me.

**Comment by sebersole** [comment]
I ended up doing both.  

I added 3 `AttributeConverter` impls:
* `org.hibernate.type.YesNoConverter`
* `org.hibernate.type.TrueFalseConverter`
* `org.hibernate.type.NumericBooleanConverter`

And an annotation form taking an enum with these 3 options:

```
@Basic
@BooleanMapping( style = Y_N )
boolean annotatedYesNo;

@Basic
@BooleanMapping( style = T_F )
boolean annotatedTrueFalse;

@Basic
@BooleanMapping( style = NUMERIC )
boolean annotatedNumeric;
```

This is all covered in the revamped basic-type section/chapter

**Reply by gavinking** [reply]
Y'know what I think would be more elegant here: letting you associate converters with annotations (via a meta-annotation).

So you could write:

```java
@Basic
@YesNo
boolean annotatedYesNo;
```

To apply the `YesNoConverter` to that field.

But of course it would also let you do a _lot_ more than that.

The reason I say this, is that I speculate the reason you introduced `@BooleanMapping` is that you dislike writing `@Convert(converter=YesNoConverter.class)`. (I don't like it much either.)

This is something I've had in the back of my mind for a long time, but it didn't immediately occur to me in this context.

**Reply by beikov** [reply]
We could use/support the stereotype annotation of CDI (https://docs.jboss.org/cdi/api/1.0/javax/enterprise/inject/Stereotype.html) for this purpose. That would be nice :)

**Reply by sebersole** [reply]
*If* this would work:

```
@Convert(converter=YesNoConverter.class)
@interface YesNo {
}
```

Then yes, this is a no-brainer.  In fact this is what I really wanted to do.  But that does not work.  

So my solution is really a practical question of having to look for 1 annotation or 3

**Reply by sebersole** [reply]
> We could use/support the stereotype annotation of CDI (https://docs.jboss.org/cdi/api/1.0/javax/enterprise/inject/Stereotype.html) for this purpose. That would be nice :)

Do you have an example of how stereotype would work here?  Not seeing it.

I specifically mean in terms of how we would handle that in terms of interpretation.  There are many ways to put lipstick on a pig, but fundamentally if the interpreter has to do extra stuff, then imo these solutions are really not any different from one another

**Reply by sebersole** [reply]
> The reason I say this, is that I speculate the reason you introduced @BooleanMapping is that you dislike writing @Convert(converter=YesNoConverter.class). (I don't like it much either.)

Yes.  Specifically the "discovery" aspect - knowing that it is available.  When mapping, users are already looking at the `annotations` package most likely.

But yes, it is verbose also

**Reply by sebersole** [reply]
To clarify, I know this "works"

```
@Convert(converter=YesNoConverter.class)
@interface YesNo {
}

... 

@Basic
@YesNo
boolean annotatedYesNo;
```

It just does not do anything.  We have to do something special to make that happen.  Specifically we have to recognize that annotation and understand that it implies a conversion.

If `@Convert` were inherited and allowed on annotaton-type then I'd be all over this (again, it was my first thought)

**Reply by gavinking** [reply]
Steve I have not thought it completely through yet (it was something I was thinking of proposing for JPA). I will think about it a bit more and reply later tonight.

**Reply by sebersole** [reply]
Now, I'm all for "fixing" JPA to allow those annotations to be composable (sorry I *hate* the phrase meta here)...

But that's a different discussion

**Reply by gavinking** [reply]
(But yeah, it would imply recognizing converter annotations at scanning time, just like what we do for `@Entity`s.)

**Reply by sebersole** [reply]
> (But yeah, it would imply recognizing converter annotations at scanning time, just like what we do for `@Entity`s.)

Which is another reason to get moved over to Jandex.. it allows that pretty easily as a built in feature.  

We could try to bake that into HCANN as an option

**Reply by beikov** [reply]
I don't know how the annotation reading works right now, but if we go through all annotations already anyway, it should be easy to inspect transitive annotations (i.e. the annotations on `YesNo`) in case an annotation is annotated with `@Stereotype`

**Reply by sebersole** [reply]
> I don't know how the annotation reading works right now, but if we go through all annotations already anyway, it should be easy to inspect transitive annotations (i.e. the annotations on `YesNo`) in case an annotation is annotated with `@Stereotype`

Overall, I am not arguing against specialized conversion markers.  I think its a great idea.  It just comes down to how that "looks" and what we have to do on the back-end.  

Here still we have to DoSomething to make that happen.  Unlike if, say, `@Convert` happened to be composable

Can you show me an example of what you think this `@YesNo`-as-a-stereotype would look like?  With or without `@Stereotype`, the problem is that you simply cannot add `@Convert`in any way-shape-form on an annotation (its `@Target` list does not allow `ANNOTATION_TYPE` nor is it inherited)

**Reply by sebersole** [reply]
Just a spit-ball...

`@Convert` cannot be composed, but we could define a "parallel":

```
@Inherited
@Target({METHOD, FIELD,ANNOTATION_TYPE})
@Retention(RUNTIME)
@interface Convert {
    <? extends AttributeConverter<?,?>> converter();
}
```

which would allow:

```
@Inherited
@Target({METHOD, FIELD,ANNOTATION_TYPE})
@Retention(RUNTIME)
@Convert( converter = YesNoConverter.class )
@interface YesNo {
}

...

@Basic
@YesNo
boolean yesOrNo;
```

**Reply by beikov** [reply]
AFAICS `javax.persistence.Convert` defines `@Target({METHOD, FIELD, TYPE})` which means it can be annotated on an annotation type as well, so no need for a new `Convert` annotation. Anyway, the stereotype example you provided here looks great. In CDI one would usually also annotate `@Stereotype` to signal that the annotation is _just a container for multiple other annotations_, but I guess we don't need to have this as long as we only support one level of depth for such stereotypes.

**Reply by sebersole** [reply]
Strange.  So `ANNOTATION_TYPE` is implied in `TYPE`.  What an odd model

Regardless, it is more the `@Inherited` that is the problem here.  Because otherwise you have to know all the kinds of annotations to look for.

**Reply by gavinking** [reply]
Hah, crazy! I think somehow I didn't know that, though it's right there in the Javadoc of a thing I've looked at hundreds of times!

**Reply by gavinking** [reply]
@beikov `@Stereotype` would make sense I guess if we were planning to combine lots of meta-annotations together in one annotation. 

But I don't think we need to go that far (yet). I can't think off the top of my head of any ORM mapping "sterotypes", i.e. annotations that I typically use in combination. Well, I _can_ think of one combination&mdash;`@Id @GeneratedValue(strategy=)`&mdash;but beyond that I'm struggling to think of anything.

Other than that, it seems to me that if we can just let you stick `@Convert` on an annotation, then that's pretty much perfect and we should just allow it.

**Reply by gavinking** [reply]
Well, arguing against myself. Here's a usecase:

```java
@Target({METHOD, FIELD})
@Retention(RUNTIME)
@Id
@GneneratedUUID
@Convert( converter = UUIDToString.class )
@interface Uuid {}
```

which combines:

- the `@Id` declaration,
- a converter, and
- a generation strategy, along the lines of what I'm proposing [here](https://github.com/eclipse-ee4j/jpa-api/pull/319#issuecomment-833376430).

Still, I would leave this idea for later.

**Reply by sebersole** [reply]
Given we are here because of sins of the past, can we please propose new annotations as composable...

```
@Inherited
@Target({METHOD, FIELD, ANNOTATION_TYPE})
@Retention(RUNTIME)
@Id
@GneneratedUUID
@Convert( converter = UUIDToString.class )
@interface Uuid {}
```

**Reply by sebersole** [reply]
Anyway, I rolled back the specific annotations here.

They, along with the static reference issues we've discussed a few times, were causing failures on Oracle because the overridden BOOLEAN JavaTypeDescriptor was not being used.

We can come back to this later

**Reply by beikov** [reply]
One common aspect that maybe even the JPA spec should care to model is mutability of types. Not sure if this is best modeled in the `AttributeConverter` contract or as annotations, but at least Hibernate supports a `@Mutability` annotation to capture this, which could also end up on such stereotype annotations. Another annotation that comes to mind is the `@JdbcType`/`@JdbcTypeCode` which could be used to define the default JDBC type code to use when a converter is applied. Again, that could end up on a stereotype annotation.

**Comment by gavinking** [comment]
@sebersole can this one be closed now?

**Reply by beikov** [reply]
I'll close this for now as I also had the impression that we decided to stick with the converter approach. If that is not correct, I'm sorry, but we can easily reopen.
